pkg/nodeidentity/aws/identify.go

```diff
@@ -102,6 +102,11 @@ func (i *nodeIdentifier) IdentifyNode(ctx context.Context, node *corev1.Node) (*
 		return nil, fmt.Errorf("found instance %q, but state is %q", instanceID, instanceState)
 	}
 
+	spot := false
+	if instance.InstanceLifecycle != nil {
```

Comment:
Shouldn't we also check that `*instance.InstanceLifecycle == "spot"`?

Reply:
> Shouldn't we also check that `*instance.InstanceLifecycle == "spot"`?

I don't think so. The field `InstanceLifecycle` is available only in spot instances, onDemand instances do not have this field.

Reply:
https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_Instance.html states that this field can have two values: `spot` and `scheduled`

Reply:
> https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_Instance.html states that this field can have two values: `spot` and `scheduled`

I missed that, thanks for bringing this to my attention! 

Given that there are two values there, I believe makes sense to revert to `string` type and pass the type of instance as an argument.

Instead of `node-role.kubernetes.io/spot-worker` maybe we should go with `node-role.kubernetes.io/instance-lifecycle: <spot|scheduled>`.

What do you think?

cc @hakman

Reply:
The `node-role.kubernetes.io/spot-worker` label is exposed through kubectl:

```
$ kubectl get nodes
NAME                            STATUS   ROLES              AGE    VERSION
ip-172-20-33-222.ec2.internal   Ready    master             147m   v1.17.5
ip-172-20-36-39.ec2.internal    Ready    node,spot-worker   125m   v1.17.5
ip-172-20-47-204.ec2.internal   Ready    node,spot-worker   123m   v1.17.5
```
Having that report `node,instance-lifecycle` would be less useful.

Reply:
I think a string type `InstanceLifecycle` would work well. Then we could also add a `node-role.kubernetes.io/scheduled` annotation.

Reply:
I think I saw mentions of `node-role.kubernetes.io/worker` label for the worker nodes, not "scheduled-worker". I did not mention this because I don't see much value in having an extra label for regular worker nodes. I don't mind if it exists though.

https://github.com/pusher/k8s-spot-rescheduler/blob/master/README.md#requirements

Reply:
"scheduled-worker" would be for scheduled instances, not onDemand.

Reply:
I updated the implementation according to @johngmyers suggestions. AWS has three types of instances:

1. OnDemand: no `InstanceLifecycle` field, no label attached
2. Spot: `InstanceLifecycle` field is set to `spot`, label attached is `node-role.kubernetes.io/spot-worker: true`
3. Scheduled: `InstanceLifecycle` field set to `scheduled` label attached is `node-role.kubernetes.io/scheduled-worker: true`

The current approach IMO is complete in comparison to previous ones as it covers all cases.

Reply:
Never knew about the scheduled ones. Thanks :).